### PR TITLE
Add missing java api for StreamTestKit

### LINKED ISSUE #970: Feature request:Add missing java api for `StreamTestKit`
Motivation:
Currently there are some methods are missing in the StreamTestkit for javadsl. And which is very annoying when you implement  a `GraphStage` in Java and test it in pure Java.

I think this task is easy if anyone else in the community want to take a try. otherwise I would like to prepare this this weekend, because I coding in java everyday.

Result:
More friendly for Java end users.

### PR DESCRIPTION
**Pull request by naosense**
To resolve #970

### CODE REVIEW

File: stream-testkit/src/main/scala/org/apache/pekko/stream/testkit/StreamTestKit.scala [hunk 1]
```diff
@@ -818,11 +886,35 @@ object TestSubscriber {
...
+     * Execute code block while bounding its execution time between `min` and
+     * `max`. `within` blocks may be nested. All methods in this trait which
+     * take maximum wait times are available in a version which implicitly uses
+     * the remaining time governed by the innermost enclosing `within` block.
+     *
+     * Note that the timeout is scaled using Duration.dilated, which uses the
+     * configuration entry "pekko.test.timefactor", while the min Duration is not.
+     *
+     * {{{
+     * val ret = within(50 millis) {
```
Comment (He-Pin) [inline stream-testkit/src/main/scala/org/apache/pekko/stream/testkit/StreamTestKit.scala:899]: delete or rewrite this code as java:)

File: stream-testkit/src/main/scala/org/apache/pekko/stream/testkit/StreamTestKit.scala [hunk 2]
```diff
@@ -818,11 +886,35 @@ object TestSubscriber {
...
+     *   test ! "ping"
+     *   expectMsgClass(classOf[String])
+     * }
+     * }}}
+     */
+    def within[T](min: java.time.Duration, max: java.time.Duration)(f: => T): T =
```
Comment (He-Pin) [inline stream-testkit/src/main/scala/org/apache/pekko/stream/testkit/StreamTestKit.scala:905]: `(f: => T)` should be `f( function.Function<T>)`

File: stream-testkit/src/main/scala/org/apache/pekko/stream/testkit/StreamTestKit.scala [hunk 3]
```diff
@@ -818,11 +886,35 @@ object TestSubscriber {
...
+
     /**
      * Same as calling `within(0 seconds, max)(f)`.
      */
     def within[T](max: FiniteDuration)(f: => T): T = probe.within(max)(f)
 
+    /**
+     * Same as calling `within(0 seconds, max)(f)`.
+     */
+    def within[T](max: java.time.Duration)(f: => T): T = probe.within(max.asScala)(f)
```
Comment (He-Pin) [inline stream-testkit/src/main/scala/org/apache/pekko/stream/testkit/StreamTestKit.scala:916]: f => T should be `function.Function[Unit, T]`

**Comment by He-Pin** [comment]
Thanks: here is some of my feedback:
1. no curried method in Java dsl
2. use the functions in `japi.function.Function` instead.
3. If you can write some spec to tests this api, then what would be nice too.

**Review by pjfanning — CHANGES_REQUESTED** [review]
looks like a good start

we need the function type changes that @kerr suggested

but we should also really have some basic test coverage of the new functions too

and all new public methods need `@since 1.1.0` in their scaladoc

**Comment by naosense** [comment]
> 3. If you can write some spec to tests this api, then what would be nice too.

@He-Pin there is no java testcase?  
image

**Comment by He-Pin** [comment]
> > 3. If you can write some spec to tests this api, then what would be nice too.
> 
> @He-Pin there is no java testcase?  
> image
> 

If you not sure how to write the test cases, I can do a following pr for your pr with tests.

**Comment by naosense** [comment]
> If you not sure how to write the test cases, I can do a following pr for your pr with tests.

yes please, master pin @He-Pin

**Comment by pjfanning** [comment]
@naosense @kerr any interest in finishing this? - we might still fit this into 1.1.0-M1 if it can be finished soon

**Comment by He-Pin** [comment]
> @naosense @kerr any interest in finishing this? - we might still fit this into 1.1.0-M1 if it can be finished soon

I think I can took this on weekend.

**Pull request closed without merging**